storage/innobase/handler/handler0alter.cc

```diff
@@ -654,9 +654,20 @@ static MY_ATTRIBUTE((warn_unused_result)) bool innobase_need_rebuild(
   Alter_inplace_info::HA_ALTER_FLAGS alter_inplace_flags =
       ha_alter_info->handler_flags & ~(INNOBASE_INPLACE_IGNORE);
 
-  if (Encryption::none_explicitly_specified(
-          ha_alter_info->create_info->explicit_encryption,
-          ha_alter_info->create_info->encrypt_type.str) ||
+  const bool was_none_explicitly_specified{
+      Encryption::none_explicitly_specified(old_table->s->explicit_encryption,
+                                            old_table->s->encrypt_type.str)};
+
+  const bool is_none_explicitly_specified{Encryption::none_explicitly_specified(
+      ha_alter_info->create_info->explicit_encryption,
+      ha_alter_info->create_info->encrypt_type.str)};
+
+  const bool is_file_per_table{
+      !tablespace_is_shared_space(ha_alter_info->create_info) &&
+      !old_table->s->system};
```

Comment:
Looks much better. Why do we need old_table->s->system check?

Reply:
I have removed it, as instructed on slack:
@robert.golebiowski patch looks good. Just one doubt on old_table->s->system. AFAIU this is TABLE_SHARE.. why do we need it?
6:35
I dont see why it is relavant here.. it is not related to real system tabelspace or anything
6:35
afaiu, it is meant for optimizer
6:35
for one-row tables.. which are also called 'system'..
6:35
// Mark 'system' tables (tables with one row) to help the Optimizer.
  share->system =
      ((share->max_rows == 1) && (share->min_rows == 1) && (share->keys == 0));

Thanks for checking this.